ros/src/waypoint_updater/waypnt_updater.py

```diff
@@ -91,69 +159,396 @@ def loop(self):
+                        self.f_wpdeq[l_ptr].state = self.state
+                        self.JMTcoeffs = None
+                        self.JMT_start_ptr = self.f_wpdeq[l_ptr].ptr_id
+                        self.JMT_time = 0.0
+                        self.JMT_max_time = 0.0
+
+                    else:
+                        self.f_wpdeq[l_ptr].waypoint.twist.twist.linear.x = \
+                            min(self.f_wpdeq[l_ptr].V, self.f_wpdeq[l_ptr].max_V)
+
+                if self.state == 'startspeedup':
+                    # dist = self.distance(last_wpd.ptr_id, self.f_wpdeq[l_ptr].ptr_id)
+                    self.f_wpdeq[l_ptr].A = self.default_accel
+                    v = max(0.5 , math.sqrt(self.f_wpdeq[l_ptr].A * disp * 2.0))
+                    v = min(v, self.f_wpdeq[l_ptr].max_V)
+                    self.f_wpdeq[l_ptr].V = v
+                    self.f_wpdeq[l_ptr].waypoint.twist.twist.linear.x = v
+
+                last_wpd = self.f_wpdeq[l_ptr]
+                rospy.loginfo("state = {}, ptr= {}, v= {}, a= {}, dist= {}".
+                            format(self.state, last_wpd.ptr_id, last_wpd.V,
+                                    last_wpd.A, disp))
             else:
-                self.default_velocity = config['dyn_default_velocity']
+                disp = self.distance(self.f_wpdeq[l_ptr].phase_start_ptr, self.f_wpdeq[l_ptr].ptr_id)
+                last_wpd = self.f_wpdeq[l_ptr]
+                rospy.loginfo("skipping recalculation "\
+                              "state = {}, ptr= {}, v= {}, a= {}, dist= {}".
+                              format(self.state, last_wpd.ptr_id, last_wpd.V,
+                                     last_wpd.A, disp))
+                
+
+    # check for state transitions
+    def state_spinner(self, last_dwp_ptr, dwp_ptr):
+
+        calc = True  # set to know if v has to be recalculated
+
+        last_deq_waypoint = self.f_wpdeq[last_dwp_ptr]
+        deq_waypoint = self.f_wpdeq[dwp_ptr]
+        if deq_waypoint.state is None:
+            if self.state not in ['JMTslowdown', 'stopped']:
+
+                if self.next_traffic_light_wp == -1:
+                    dist = 1000  # far away
+                else:
+                    dist = self.distance(deq_waypoint.ptr_id,
+                                         self.next_traffic_light_wp)
+
+                if dist < self.get_stopping_distance(last_deq_waypoint.V):
+                    self.state = 'JMTslowdown'
+                    # setup JMT for calcs
+
+                    end_wp = self.next_traffic_light_wp - 1
+                    dist = self.distance(deq_waypoint.ptr_id, end_wp)
+                    # try to bias to slowing down from start by setting A slightly negative
+                    start = list([0, last_deq_waypoint.V, last_deq_waypoint.A-0.1])
+                    end = list([dist, 0.0, 0.0])
+                    time = 2.0 * 2 * dist / last_deq_waypoint.V
```

Comment:
Is this recalculating the JMT coefficients at every predicted time for each waypoint all the way to the final state?